README.md
# baffle
Waffle graphs in base R graphics.

## Installation

You can install the latest version of baffle from [CRAN](https://CRAN.R-project.org) with:

```r
install.packages("baffle")
```

or directly from GitHub using the `devtools` package:
```r
devtools::install_github("j-moravec/baffle")
```

## Example

See [online documentation](https://j-moravec.github.io/baffle/) for a more complex examples.

```r
library(baffle)
waffle(c(25,75), col=c("darkorchid", "lightgray"))
```

![](man/figures/waffle.svg)


NEWS.md
## Version 0.2.1
* Added functions for drawing rounded corners.

## Version 0.2.0
* Completely reworked to increase flexibility.
* Added a number of shape functions for different feel and look of waffles.
* Added support for stacked and unstacked waffles.

## Version 0.1.0
* basic proof of concept


tests/testthat.r
library(testthat)
library(baffle)

test_check("baffle")


tests/testthat/test-design.r
square.matrix = function(x, n){
    matrix(as.integer(x), n, n, byrow=TRUE)
    }
        
test_that("Design matrix is filled in the correct direction", {
    x = c(2,0,1)
    
    expect_identical(design(x), square.matrix(c(3, NA, 1, 1), 2))
    expect_identical(design(x, byrow=FALSE), square.matrix(c(1, NA, 1, 3), 2))
    expect_identical(design(x, from="topleft"), square.matrix(c(1, 1, 3, NA), 2))
    expect_identical(design(x, from="bottomright"), square.matrix(c(NA, 3, 1, 1), 2))
    expect_identical(design(x, byrow=FALSE, from="topright"),
                     square.matrix(c(3, 1, NA, 1), 2))
    })



test_that("Design matrix is constructed with the correct dimensions", {
    expect_identical(design(c(3,3,3)) |> dim(), c(3L,3L))
    expect_identical(design(c(3,3,3,1)) |> dim(), c(4L,4L))
    expect_identical(design(c(0,1,0)) |> dim(), c(1L,1L))
    
    expect_identical(design(0, nrow=3, ncol=3) |> dim(), c(3L,3L))
    expect_error(design(0), "Unable to determine dimension of the matrix.")
    })


test_that("Unstacked design matrix is constructed correctly", {
    expect_identical(design(c(1, 2, 3), stacked=FALSE), 
        matrix(as.integer(c(NA, 1, 2, 2, 3, 3, NA, 3)), 2, 4))
    expect_identical(design(c(1, 2, 3), stacked=FALSE, horiz=FALSE),
        matrix(as.integer(c(1, 2, 3, 3, NA, 2, NA, 3)), 4, 2))
    })


tests/testthat/test-recycle.r
test_that("Recycling dots is done correctly", {
    expect_identical(
        recycle_dots(
            x = c(1, 2, 3, 1 , 2),
            col = c("red", "green", "blue"),
            pch = 19,
            lwd = c(1, 2)
            ),
        list(
            col = c("red", "green", "blue", "red", "green"),
            pch = 19,
            lwd = c(1, 2, 1, 1, 2)
            )
        )
    })


R/raster.r
#' Plot raster image
#'
#' Plot raster image centered at `x` and `y` coordinates scaled to diameter `d`.
#'
#' The `rasters()` function is a convenient wrapper around [graphics::rasterImage()] with similar
#' interface to the [Shapes] functions available in this package (such as [square()], [circle()]
#' and [rcpoly()]).
#'
#' The raster image is plotted centered at the `x` and `y` coordinates and scaled to the diameter
#' size `d`. When `dx` and `dy` are `NA`, the proportions of the raster are kept unchanged,
#' otherwise they are scaled to the specified size in either direction. This scaling is done
#' before rotation.
#'
#' Unlike in `rasterImage`, the rotation is performed clockwise and the rotation axis is the center
#' of the raster (i.e., the provided x and y coordinates), rather than the bottom left coordinate
#' `x0`. This rotation is performed after scaling.
#'
#' As of yet, the rotation is accurate only when the aspect ratio is set to 1 (`asp=1`) through
#' the `graphics::plot.window()` call.
#'
#' The `rasters()` function is fully vectorized.
#'
#' @param x,y coordinates
#' @param image raster image
#' @param d **optional** diameter, see details
#' @param dx,dy **optional** diameter in either coordinate direction
#' @param rotate **optional** clockwise rotation in degrees (0-360°)
#' @param ... **optional** other parameters passed to [graphics::rasterImage()],
#'  such as `interpolate`
#' @return No return value, called for side effects
#'
#' @examples
#' # create plotting window
#' plot.new(); plot.window(c(-1,1), c(-1,1), asp=1); axis(1); axis(2)
#'
#' # create raster image, alpha is convenient when overplotting
#' img = matrix(adjustcolor("black", alpha.f=0.3), 3, 3)
#' img[2, 2] = adjustcolor("white", alpha.f=0.3)
#' img = as.raster(img)
#'
#' rasters(0, 0, img)
#'
#' # interpolate=FALSE makes quite a difference
#' rasters(0, 0, img, interpolate=FALSE)
#'
#' # arguments are vectorized, standard recycling rules apply
#' rasters(0, 0, img, interpolate=FALSE, rotate=c(30, 60, 90))
#' rasters(c(-1, -0.5,  0.5, 1), c(1, 0.5, -0.5, -1), img, interpolate=FALSE)
#'
#' @export
#' @seealso [grDevices::as.raster()] and [graphics::rasterImage()]
rasters = function(x, y, image, d=0.9, dx=NA, dy=NA, rotate=0, ...){
    if(grDevices::is.raster(image))
        image = list(image)

    Map(f=raster, x=x, y=y, image=image, d=d, dx=dx, dy=dy, rotate=rotate, ...)
    
    invisible(NULL)
    }

# internal 
raster = function(x, y, image, d=0.9, dx=NA, dy=NA, rotate=0, ...){

    if(is.na(dx) && is.na(dy)){
        prop = dim(image) / max(dim(image))
        dx = d * prop[1]
        dy = d * prop[2] 
        }

    if(is.na(dx))
        dx = d
    if(is.na(dy))
        dy = d

    if(rotate == 0){
        graphics::rasterImage(image, x - dx/2, y - dy/2, x + dx/2, y + dy/2, angle=0, ...)
        invisible(NULL)
        }

    origin = c(x - dx/2, y - dy/2)
    cent = .rotate(c(x, y), -rotate, origin=origin)
    
    x = x - cent[1] + x
    y = y - cent[2] + y

    graphics::rasterImage(
        image,
        x - dx/2,
        y - dy/2,
        x + dx/2,
        y + dy/2,
        angle=-rotate,
        ...
        )
    }


# internal
.rotate = function(p, angle, origin){
    angle = angle * pi / 180
    
    co = cos(angle)
    si = sin(angle)
    
    x1 = p[1] - origin[1]
    y1 = p[2] - origin[2]
    
    x2 = x1 * co - y1 * si + origin[1]
    y2 = x1 * si + y1 * co + origin[2]
    
    c(x2, y2)
    }


R/shapes.r
#' Plot convex polygons
#'
#' A collection of functions for plotting polygonal shapes.
#'
#' Polygons are drawn centered on the `x` and `y` coordinates, with a diameter `d`
#' (or `dx` and `dy`). Typically, different shapes are obtained through a parametrization
#' of the `shapes()` function, which draws a convex polygon using the [graphics::polygon()],
#' with the exception of `square()` and `rectangle()` function, which use the [graphics::rect()]
#' function instead, behave slightly differently, and should be slightly faster, and the
#' `rounded_square()` and `rounded_rectangle()`, which use `round_rect()` function.
#'
#' The diameter parameter `d` is interpreted differently depending for `square()` and `rectangle()`
#' and for other polygonal functions build on the `shapes()` function (`circle()`, `ellipse()`
#' and `rcpoly()`). For the `square()` and `rectangle()`, the diameter is the size of the square,
#' `d=1` thus fills the whole 1x1 tile. For `shapes()` function, `d` is the diameter
#' of the inscribed circle to the square of size `d`. This is more convenient solution to prevent
#' accidental overplotting when individual shapes are plotted next to each other in regular
#' intervals, the distance between such points would be equal to the diameter in both cases.
#' See examples.
#'
#' All shapes function accept the graphical parameters `col`, `border`, `lty` and `lwd`, which
#' are passed to the [graphics::polygon()] and [graphics::rect()]. Apart of a different default
#' values, they behave in the same way.
#'
#' All parameters are vectorized and will recycle as required, with the typical warning if
#' parameters are not multiply of each other. This can be used to create pleasant geometric images.
#' See examples. 
#'
#' @param x,y coordinates
#' @param n the number of vertices of polygon, with the minimum of three (triangle). Large `n`, such
#'   as `n=1000` approximate circle. The vertices start at the 12 o'clock position and are placed
#'   clockwise in a regular intervals. For `rounded_square` and `rounded_rectangle`, the number of
#'   vertices that approximate the circular shape of the rounded corners. Small `n`, such as
#'   `n=10` should have sufficient smoothness for most applications. 
#' @param d **optional** diameter, see details
#' @param dx,dy **optional** diameter in either coordinate direction
#' @param r,rx,ry **optional** for rounded corner, the proportion of side marking the start of
#'   rounded corner.
#' @param rotate **optional** clockwise rotation in degrees (0-360°),
#'   not available for `square` and `rectangle`
#' @param ... **optional** graphical parameters `col`, `border`, `lty` and `lwd` passed to
#'   [graphics::polygon()] or [graphics::rect()]
#' @return No return value, called for side effects
#' @importFrom graphics par
#'
#' @examples
#' plot(0, 0) # create plotting window
#'
#' # Following calls are equivalent
#' square(0, 0, 1)
#' rectangle(0, 0, 1)
#' rectangle(0, 0, dx=1, dy=1)
#'
#' # Not equivalent to `square`
#' rcpoly(0, 0, 4, d=1)
#'
#' # Same output as `square`, but not equivalent
#' rcpoly(0, 0, 4, d=sqrt(2), rotate=45)
#'
#' # Vectorizing parameters
#' plot(0, 0)
#' rotate = seq(0, 18, by=30)
#' d = seq(1, by=-0.1, length.out = length(rotate))
#' rcpoly(0,0,3, border="red", lwd=3, rotate=rotate, d=d)
#'
#' @seealso [graphics::polygon()] and [graphics::rect()] for the underlying plotting functions
#' @name Shapes
NULL

#' @describeIn Shapes draw squares
#' @export
square = function(x, y, d=0.9, ...){
    .rectangle(x=x, y=y, d=d, ...)
    }

#' @describeIn Shapes draw rectangles
#' @export
rectangle = function(x, y, d=0.9, dx=d, dy=d, ...){
    .rectangle(x=x, y=y, d=d, dx=dx, dy=dy, ...)
    }


# internal
.rectangle = function(
    x, y, d=0.9, dx=d, dy=d,
    col=par("fg"), border=NA, lty=par("lty"), lwd=par("lwd")
    ){
    graphics::rect(x - dx/2, y - dy/2, x + dx/2, y + dy/2,
                   col=col, border=border, lty=lty, lwd=lwd)
    }


#' @describeIn Shapes draw squares with rounded corners
#' @export
rounded_square = function(x, y, d=0.9, r=0.2, n=10, ...){
    .rounded_rectangle(x=x, y=y, d=d, r=r, n=n, ...)
    }


#' @describeIn Shapes draw rectangles with rounded corners
#' @export
rounded_rectangle = function(
    x, y, d=0.9, dx=d, dy=d,
    r=0.2, rx=r, ry=r, n=10,
    ...
    ){
    .rounded_rectangle(x=x, y=y, d=d, dx=dx, dy=dy, r=r, rx=rx, ry=ry, n=n, ...)    
    }


# internal
.rounded_rectangle = function(
    x, y, d=0.9, dx=d, dy=d, r=0.2, rx=r, ry=r, n=10,
    col=par("fg"), border=NA, lty=par("lty"), lwd=par("lwd"),
    ...
    ){
    round_rect(x - dx/2, y - dy/2, x + dx/2, y + dy/2, xr=rx, yr=ry, n=n,
               col=col, border=border, lty=lty, lwd=lwd, ...)
    }

#' @describeIn Shapes draw circles
#' @export
circle = function(x, y, d=0.9, n=1000, ...){
    shapes(x=x, y=y, n=n, d=d, ...)
    }

#' @describeIn Shapes draw ellipses
#' @export
ellipse = function(x, y, d=0.9, dx=d, dy=d, n=1000, ...){
    shapes(x=x, y=y, n=n, d=d, dx=dx, dy=dy, ...)
    }

#' @describeIn Shapes draw regular convex polygons
#' @export
rcpoly = function(x, y, n, d=0.9, rotate=0, ...){
    shapes(x=x, y=y, n=n, d=d, rotate=rotate, ...)
    }


# internal
shape = function(
    x, y, n, d=0.9, dx=d, dy=d, rotate=0,
    col=par("fg"), border=NA, lty=par("lty"), lwd=par("lwd")
    ){

    points = seq(0, 2*pi, length.out=n+1)
    rotate = (pi / 180) * rotate

    graphics::polygon(
        sin(points + rotate)/2*dx + x,
        cos(points + rotate)/2*dy + y,
        col=col, border=border, lty=lty, lwd=lwd
        )
    }


#' @describeIn Shapes draw convex polygons
#' @export
shapes = function(x, y, n, d=0.9, dx=d, dy=d, rotate=0, ...){

    if(any(n < 3))
        stop("The minimum number of vertices must be 3")

    Map(f=shape, x=x, y=y, n=n, d=d, dx=dx, dy=dy, rotate=rotate, ...)

    invisible(NULL)
    }


R/round_rect.r
#' Plot a rectangle with rounded corners
#'
#' Plot one or more rectangles with rounded corners.
#'
#' The rounded corner is a part of a circle (one quarter) drawn between two points of neighbouring
#' sides of a rectangle. The relative position of these two points determine how rounded will the
#' final shape be. The position of these points is determined by the parameters `xr` and `yr`,
#' which determine the proportion of side x and y, from which the rounded corner is drawn.
#' Values between 0 and 1 are permitted, but given the symmetricity of a rectangle, values  larger
#' than 0.5 are reflected back (modulo 0.5). When both xr and yr are 0 or 1, normal rectangle
#' without rounded corners is drawn. When xr and yr are 0.5, ellipsis is drawn.
#'
#' @param x1,y1 x and y coordinates of the bottom left corner
#' @param x2,y2 x and y coordinates of the top right corner
#' @param xr,yr the proportion of length of a side of the rectangle, from which the rounded corners
#'              start. Values between 0 and 1 are permitted. See details.
#' @param n the number of points used to approximate the curvature of the rounded corner
#' @param ... other parameters passed to [graphics::polygon()].
#' @return No return value, called for side effects
#'
#' @export
round_rect = function(x1, y1, x2, y2, xr=0.2, yr=0.2, n=10, ...){
    Map(f=.round_rect, x1=x1, y1=y1, x2=x2, y2=y2, xr=xr, yr=yr, n=n, ...)
    invisible(NULL)
    }

# internal
.round_rect = function(
    x1, y1, x2, y2, xr=0.2, yr=0.2, n=10,
    col=NA, border=NULL, lty=par("lty"), lwd=par("lwd"),
    ...
    ){
    stopifnot(xr > 0 && xr < 1)
    stopifnot(yr > 0 && yr < 1)

    xr = xr %% 0.5
    yr = yr %% 0.5

    xs = x2 - x1
    ys = y2 - y1

    curve = seq(0, pi/2, length.out=n) |> sin()

    x = c(x1, rev(1-curve)*xs*xr+x1, xs*xr+x1,
          xs*(1-xr)+x1, curve*xs*xr + xs*(1-xr)+x1, x2)
    y = c(ys*yr+y1, (1-curve)*ys*yr+y1, y1,
          y1, rev(1-curve)*ys*yr+y1, ys*yr+y1)

    graphics::polygon(
        x = c(x, x2, rev(x), x1),
        y = c(y, ys*(1-yr)+y1, y1 + y2 - rev(y), ys*yr+y1),
        col = col, border = border, lty = lty, lwd = lwd,
        ...
        )
    }


R/waffle.r
#' Make a waffle chart.
#'
#' Waffle chart is a pie chart that visually represents abundances with the number of squares.
#' The waffle chart consists of squares plotted on a rectangular lattice according to a design
#' matrix that is constructed from a vector of abundances or can be provided directly.
#'
#' @details
#' The `waffle()` function accepts a vector of abundances and plots a waffle chart.
#' This is done by first constructing a design matrix using the `design()` function,
#' which is then parsed to the `waffle.mat()` function.
#'
#' The `design()` functions construct a design matrix according.
#' The design matrix is filled by an integer vector derived from the abundance vector `x`.
#' Each integer correspond to the order of abundances in `x`, with the quantity equal to the value
#' in `x`. This means that for the `x=c(3,5)`, the design matrix will be filled with three `1`
#' and five `2`, all other cells of the design matrix are set as unknown values `NA`.
#'
#' By default, the design matrix is filled by row, starting from the bottom left corner,
#' this can be changed by setting the variables `byrow` and `from`.
#' By setting `byrow=FALSE`, the matrix is filled by columns first.
#'
#' If `ncol` or `nrow` is not specified, a squared matrix that will fit the sum of abundances
#' will be constructed.
#'
#' By default, the design matrix is stacked, i.e., all the elements are placed in a single
#' matrix object, similarly how barplot produces stacked lines. By specifying `stack=FALSE`,
#' the `design()` unstacks the elements, each are then placed in their own matricies, which
#' are then connected by `NA` elements of size `gap`. More complex 
#'
#' If the input vector `x` is a named vector, the names are preserved in the `levels`
#' attribute. These levels are not currently used, but might be used in the future
#' for automatic legend creation and subsetting.
#'
#' The function `waffle.mat()` accepts a custom-made design matrix and thus allows a better
#' control of colored regions. It is called internally by the `waffle()` function, which serves
#' as an easy to use interface for the `waffle.mat()`. For this reason, the `waffle.mat()`
#' does not checks for the validity of input arguments and does not set a default colors.
#'
#' The assumed and allocated coordinate system is from `0` to `ncol` for x and
#' from `0` to `ncol` for y. Squares are filled from top right corner of this coordinate
#' system.
#'
#' If `add=FALSE`, a new window with a fixed aspect ratio x/y=1 is allocated so that
#' plotted polygons are squares (by default). This might cause the plot margins,
#' and thus the main title, to be quite far away. In this case, plotting the title using `text()`
#' instead of `title()` might be a better idea. In this case, the coordinates might be:
#' \code{text(x=(ncol+2)/2, y=nrow+1,...)}
#'
#'
#'
#' @param x a vector of abundances or a design matrix (see details)
#' @param nrow **optional** the number of rows
#' @param ncol **optional** the number of columns
#' @param byrow **optional** fill matrix by rows or by columns
#' @param from **optional** starting position from which the matrix is filled, one of
#'   "bottomleft", "bottomright", "topleft" or "topright".
#' @param stacked **optional** if FALSE, produce an unstacked waffle chart, see details
#' @param gap **optional** gap between unstacked subwaffles, works only for unstacked waffle charts
#' @param horiz **optional** if FALSE, vertical instead of horizontal waffle chart is produced,
#'   works only for unstacked waffle charts
#' @param add **optional** whether to add to a current plot
#' @param f **optional** a shape function (see details)
#' @param ... **optional** other parameters passed to the `f` function
#' @return `design()` returns a design matrix (see details), in addition, if `x` is a named vector,
#'   the names are preserved in the `levels` attribute.
#'   `waffle()` and `waffle.mat()` do not have a return value
#'
#' @seealso the `waffle` package for a `ggplot2` version.
#'
#' @examples
#' waffle(c(50,25,25))
#' waffle(c(25,75), col=c("darkorchid", "lightgray"))
#' waffle(c(14,8,4), nrow=3)
#'
#' # custom design matrix with a more complex structure
#' cols = palette.colors(3, "Set 1")
#' design_mat = matrix(NA, 7,10)
#' design_mat[1,] = 1
#' design_mat[2,1:8] = 1
#' design_mat[4,] = 2
#' design_mat[5,1:4] = 2
#' design_mat[7,1:5] = 3
#' waffle.mat(design_mat, col=cols)
#' @export
waffle = function(
    x, f=NULL, ...,
    nrow=NULL, ncol=NULL, byrow=TRUE, from="bottomleft",
    stacked=TRUE, gap=0, horiz=TRUE, add=FALSE
    ){

    # construct design matrix
    mat = design(
        x, nrow=nrow, ncol=ncol, byrow=byrow, from=from,
        stacked=stacked, gap=gap, horiz=horiz
        )

    # plot the waffle using the design matrix
    waffle.mat(mat, f=f, ..., add=add)
    }


#' @rdname waffle
#' @export
waffle.mat = function(x, f=square, ..., add=FALSE){
    # TODO can we utilize the levels() from [design]? Perhaps to make automatic legend?

    nrow = nrow(x)
    ncol = ncol(x)
    x = as.vector(x)
    pick = !is.na(x)
    x = x[pick]

    if(is.null(f))
        f = square

    dots = list(...)

    n = max(x)
    if(is.null(dots[["col"]]) && n < 10)
        dots[["col"]] = grDevices::palette.colors(n, "Set 1")

    if(is.null(dots[["col"]]) && n >= 10)
        dots[["col"]] = grDevices::hcl.colors(n, "Zissou 1")

    dots = do.call(recycle_dots, c("x"=list(x), dots))

    xx = matrix(1:ncol-1, nrow, ncol, byrow=TRUE)
    xx = as.vector(xx)[pick] + 0.5

    yy = matrix(nrow:1-1, nrow, ncol)
    yy = as.vector(yy)[pick] + 0.5

    if(!add){
        graphics::plot.new()
        graphics::plot.window(xlim=c(0, ncol), ylim=c(0, nrow), asp=1)
        }


    # vectorize over vector of functions f
    if(length(f) > 1){
        f = rep(f, length.out=n)
        f = f[x]
        
        h = function(g, ...){do.call(g, list(...))}
        .mapply(h, dots=c("g"=list(f), "x"=list(xx), "y"=list(yy), dots), NULL)

        return(invisible())
        }

    do.call(f, args=c("x"=list(xx), "y"=list(yy), dots))
    }



#' Recycle dots args
#'
#' @param x an integer vector, typically with values ranging from 1 to n, with each integer
#'   with each integer specifying an index of element for each item in `...`
#' @param ... additional arguments, if length of item of `...` is larger than one, its elements are
#'   is recycled according to indiced specified by the vector `x`
#' @return recycled arguments of `...`
#' @keywords internal
recycle_dots = function(x, ...){
    recycle = function(x, vector, length){
        if(length(x) > 1){
            x = rep(x, length.out=length)
            x = x[vector]
            }

        x
        }

    n = max(x) # TODO integrate with levels later

    dots = list(...)

    dots[] = lapply(dots, recycle, vector=x, length=n)

    dots
    }

#' @rdname waffle
#' @export
design = function(
    x, nrow=NULL, ncol=NULL, byrow=TRUE, from="bottomleft",
    stacked=TRUE, horiz=TRUE, gap=0
    ){
    if(stacked){
        design_stacked(x, nrow=nrow, ncol=ncol, byrow=byrow, from=from)
        } else {
        design_unstacked(x, nrow=nrow, ncol=ncol, byrow=byrow, from=from, horiz=horiz, gap=gap)
        }
    }


# internal
design_stacked = function(x, nrow=NULL, ncol=NULL, byrow=TRUE, from="bottomleft"){
    from = match.arg(from, c("bottomleft", "bottomright", "topleft", "topright"))

    total = sum(x)

    if(any(x < 0))
        stop("Elements of `x` need to be positive.")
    if(any(as.integer(x) != x))
        stop("`x` must be integer.")
    if(total == 0 && (is.null(nrow) || is.null(ncol) ) )
        stop("Unable to determine dimension of the matrix. The sum of `x` is zero.")

    if(is.null(nrow) && is.null(ncol))
        nrow = ncol = ceiling(sqrt(total))
    if(is.null(nrow))
        nrow = ceiling(total / ncol)
    if(is.null(ncol))
        ncol = ceiling(total / nrow)

    cells = nrow*ncol

    if(total > cells)
        stop("The sum of elements in `x` must be smaller or equal to the total number of cells")

    mat = rep(seq_along(x), x)
    mat = c(mat, rep(NA, cells - total))
    mat = matrix(mat, nrow, ncol, byrow=byrow)

    # by default, matrix() is filled from top left corner, so we use some switches to flip it
    if(from %in% c("bottomleft", "bottomright"))
        mat = mat[nrow:1, , drop=FALSE]

    if(from %in% c("bottomright", "topright"))
        mat = mat[, ncol:1, drop=FALSE]

    if(! names(x) |> is.null() )
        levels(mat) = names(x)

    mat
    }


# internal
design_unstacked = function(
    x, nrow=NULL, ncol=NULL, byrow=TRUE, from="bottomleft", gap=0, horiz=TRUE
    ){
    if(is.null(nrow) && horiz)
        nrow = max(ceiling(sqrt(x)))

    if(is.null(ncol) && !horiz)
        ncol = max(ceiling(sqrt(x)))

    n = length(x)
    template = vector("numeric", length=n)
    mat = vector("list", length=n)
    for(i in seq_len(n)){
        y = template
        y[i] = x[i]
        
        mat[[i]] = design(y, nrow=nrow, ncol=ncol, byrow=byrow, from=from)
        }
    
    if(horiz){
        empty = matrix(NA, nrow=nrow, ncol=gap)
        mat[[n+1]] = empty
        mat = mat[interleave(n+1, seq_len(n))[-1]]
        mat = do.call(cbind, mat)
        } else {
        empty = matrix(NA, ncol=ncol, nrow=gap)
        mat[[n+1]] = empty
        mat = mat[interleave(n+1, seq_len(n))[-1]]
        mat = do.call(rbind, mat)
        }
        
    mat
    }

# internal
interleave = function(...){
    as.vector(rbind(...))
    }
